File: connect-four.rb
#! /usr/bin/ruby

# Wendy Smoak | wsmoak@gmail.com | http:/wsmoak.net | October 2014
# Connect Four game 

#initialize game board array
ROWS, COLS = 6, 7
$board = Array.new(ROWS*COLS)

$last_move = {row: nil, col: nil}

UP, DOWN, LEFT, RIGHT, SAME = -1, 1, -1, 1, 0
WIN = 4

#display the board, visualizing single array as rows and columns
def visualize_array
(0...ROWS).each { |row|
  (0...COLS).each { |col|
    print row.to_s + " " + col.to_s + " "
    print "("  +  (COLS*row + col).to_s + ")   " # index into board array
  }
  puts
}
end

def display_board
  (0...ROWS).each { |row|
    (0...COLS).each { |col|
      i = COLS*row + col
      print $board[i] ? $board[i].to_s + "  " : ".  "
    }
    puts
  }
end

def choose(player)
  done=false
  until done do

    puts "Player "+player.to_s+", choose a column (0-"+(COLS-1).to_s+")"
    col = gets.chomp.to_i

    # if row 0 for that column is filled in, prompt for a different column
    if $board[col] then puts "column is full"; redo; end

    #find the lowest empty row in that column
    #do this "upside down" so all I have to find is the first empty position.
    (ROWS-1).downto(0).each do |row|
      i = COLS*row + col
      if !$board[i] then
        $board[i] = player
        $last_move = {row: row, col: col, player: player}
        done=true
        break
      end
    end
  end
end

# returns the index into the array, or nil if the position is not on the board
def index (row, col)
  if (0...ROWS).member? row
    if (0...COLS).member? col
      COLS*row + col
    end
  end
end

# returns a hash with the row, column and array index, or nil if the position is not on the board
def position (r,c)
  if (0...ROWS).member? r
    if (0...COLS).member? c
      {row: r, col: c, pos: COLS*r + c}
    end
  end
end

def find_match(row_dir, col_dir)
  row,col = $last_move[:row], $last_move[:col]
  count = 0
  match = true
  while match do
    col += col_dir # move right or left
    row += row_dir # move up or down
    square = position(row,col)
    if square && $board[square[:pos]] == $last_move[:player] then
      count += 1
    else
      match = false
    end
  end
  return count
end

def find_row_match
  find_match(SAME,RIGHT) + find_match(SAME,LEFT)
end

def winner?(count)
  count + 1 >= WIN
end

def last_player
   $last_move[:player].to_s
end

def end_game
  puts "Winner is " + last_player
  exit
end

def check_row
  if winner? find_row_match then end_game end
end

def find_column_match
  find_match(DOWN,SAME)
end

def check_column
  puts "Checking column"
  if winner? find_column_match then end_game end
end

def find_backward_diagonal_match
  find_match(DOWN,RIGHT) + find_match(UP,LEFT)
end

def find_forward_diagonal_match
  find_match(DOWN,LEFT) + find_match(UP,RIGHT)
end

def check_diagonals
  puts "check_diagonals"
  if winner? find_backward_diagonal_match then end_game end
  if winner? find_forward_diagonal_match then end_game end
end

def check_tie
  #if we don't have a winner at this point, and all the spaces are filled (not nil), then it's a tie
  if $board.all?
    puts "It's a tie!"
    exit
  end
end

def check_board
  check_row
  check_column
  check_diagonals
  check_tie
end

def play
  loop do
    choose(:B)
    display_board
    check_board

    choose(:R)
    display_board
    check_board
  end
end

display_board
play

File: connect-four.js
// Wendy Smoak | wsmoak@gmail.com | http:/wsmoak.net | October 2014
// Connect Four game
(function() {
    // see http://benalman.com/news/2010/11/immediately-invoked-function-expression/

    var ROWS = 6;
    var COLS = 7;

    var Game = function(canvasId) {
            var red = true;

            // see http://html5doctor.com/an-introduction-to-the-canvas-2d-api/
            var canvas = document.getElementById(canvasId);
            // see http://www.w3.org/TR/2dcontext/ 
            var context = canvas.getContext('2d');

            drawColumnGrid(context, COLS);
            drawRowGrid(context, ROWS);

            canvas.addEventListener('click', function(event) {
                var rowHeight = Math.floor(context.canvas.height / ROWS);
                var columnWidth = Math.floor(context.canvas.width / COLS);

                // we don't care what row they clicked on, the game piece will drop to the lowest available row
                var col = Math.floor(event.pageX / columnWidth);

                console.log("Clicked! " + event.pageX + " " + event.pageY + " in Column " + col);

                var x = Math.floor((col * columnWidth) + (columnWidth / 2)) // x position in middle of column
                //ignore click events that fall outside the width of the canvas (how does this happen?)
                if (col >= 0 && col < COLS) {
                    //find first empty row in that column
                    for (var row = ROWS; row > 0; row--) {
                        var y = Math.ceil((row * rowHeight) - (rowHeight / 2)); // y position in middle of row
                        // see http://www.w3schools.com/tags/canvas_getimagedata.asp
                        var imageData = context.getImageData(x, y, 1, 1);
                        console.log("Image Data for " + x + " " + y + " is " + imageData.data[0] + " " + imageData.data[1] + " " + imageData.data[2] + " " + imageData.data[3]);
                        if (imageData.data[3] === 0) { //if transparent meaning the space is empty
                            drawCircle(context, x, y, red);
                            checkBoard(context, x, y, red);
                            red = !red; //toggle player color true/false -> red/black
                            break;
                        } // end if
                    } //end for
                } //end if
            });


        };

    var drawColumnGrid = function(context, cols) {
            var columnWidth = context.canvas.width / cols;
            for (var i = 0; i < cols; i++) {
                context.rect(i * columnWidth, 0, columnWidth, context.canvas.height);
                context.strokeStyle = 'yellow';
                context.lineWidth = '3';
                context.stroke();
            }
        };

    var drawRowGrid = function(context, rows) {
            var rowHeight = context.canvas.height / rows;
            for (var i = 0; i < rows; i++) {
                context.rect(0, i * rowHeight, context.canvas.width, rowHeight);
                context.strokeStyle = 'yellow';
                context.lineWidth = '3';
                context.stroke();
            }
        };

    var drawCircle = function(context, x, y, red) {
            console.log("Drawing circle!");
            //see http://www.html5canvastutorials.com/tutorials/html5-canvas-circles/
            //see http://www.tauday.com/tau-manifesto
            var TAU = 2 * Math.PI;
            context.beginPath();
            //see http://www.w3schools.com/tags/canvas_arc.asp
            context.arc(x, y, 10, 0, TAU);
            context.fillStyle = red ? 'red' : 'black';
            context.fill();
        };

        var findMatch = function(context, x, y, rowDirection, colDirection) {
		    var count = 0;
		    var lastMove = context.getImageData(x, y, 1, 1);
    	  //console.log("Last Move is " + x + " " + y + " is " + lastMove.data[0] + " " + lastMove.data[1] + " " + lastMove.data[2] + " " + lastMove.data[3]);
			xpos = x;
			ypos = y;
		    // look to the left and count matches
			while ( xpos > 0 && xpos < context.canvas.width && ypos > 0 && ypos < context.canvas.height) {
                xpos += colDirection;
				ypos += rowDirection;
		        var imageData = context.getImageData(xpos, ypos, 1, 1);
                //console.log("Image Data for " + xpos + " " + ypos + " is " + imageData.data[0] + " " + imageData.data[1] + " " + imageData.data[2] + " " + imageData.data[3]);
		        if (imageData.data[3] > 0 && //not transparent
		           lastMove.data[0] === imageData.data[0] &&
		           lastMove.data[1] === imageData.data[1] &&  //same color
		           lastMove.data[2] === imageData.data[2] ) {
			       count += 1;
		        } else {
			       break;
		        }
		    } // end while
		console.log ( "returning " + count );
			return count;
        };

	    var checkBoard = function(context, x, y, red) {
		    var columnWidth = Math.floor(context.canvas.width / COLS);
		    var rowHeight = Math.floor(context.canvas.height / ROWS);

			var LEFT = -columnWidth;
			var RIGHT = columnWidth;
			var UP = -rowHeight;
			var DOWN = rowHeight;
			var SAME = 0;

			rowWin = findMatch(context,x,y,SAME,LEFT) + findMatch(context,x,y,SAME,RIGHT);
		    console.log ( "Counted left and right matches: " + rowWin );

		    colWin = findMatch(context,x,y,DOWN,SAME);
		    console.log ( "Counted up and down matches: " + colWin );

			backDiagWin = findMatch(context,x,y,UP,LEFT) + findMatch(context,x,y,DOWN,RIGHT);
		    console.log ( "Counted back diagonal matches: " + backDiagWin );

			fwdDiagWin = findMatch(context,x,y,UP,RIGHT) + findMatch(context,x,y,DOWN,LEFT);
		    console.log ( "Counted fwd diagonal matches: " + fwdDiagWin );


        };

    window.onload = function() {
        new Game("screen");
    };

})();
